mautic/mautic
# Segment rebuild improvements

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [❌]
| New feature/enhancement? (use the a.x branch)      | [✅]
| Deprecations?                          | [❌]
| BC breaks? (use the c.x branch)        | [❌]
| Automated tests included?              | [❌] 
| Related user documentation PR URL      | 
| Related developer documentation PR URL |
| Issue(s) addressed                     | 



#### Description:



This PR introduces some improvements to segment rebuild code.

1. Added `$date` parameter to `app/bundles/LeadBundle/Event/ListChangeEvent.php` to allow triggering/processing of this event in an asynchronous way.
2. Introduced an `'unbuffered'` DB connection that allows iterating over huge DB result without any memory issues. The `'unbuffered'`connection can be retrieved from the container via `'doctrine.dbal.unbuffered_connection'`.
3. Minor refactoring of `app/bundles/LeadBundle/Segment/ContactSegmentService.php`. Some duplicate code was removed using `LeadBatchLimiterTrait`.

#### Steps to test this PR:


1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Make sure the segments are built properly.
